modules/govuk_ci/manifests/master.pp

```diff
@@ -46,6 +46,14 @@
 
   # Add pipeline jobs from applications hash in Hieradata
   create_resources(govuk_ci::job, $pipeline_jobs)
+  # Manually delete all jobs which aren't in the hash
+  # lint:ignore:single_quote_string_with_variables
+  $full_rm_command = sprintf('%s%s%s', 'key_str=$(IFS=\|; printf \'%s\' "${', keys($pipeline_jobs), '[*]}"); ls -a | grep -vE "$key_str|\.|\.\." | xargs rm -rf')
```

Comment:
This is very hard to read and even harder to test. How about making it a separate shell script? You could pass the list of directories to keep via a Puppet template.

Reply:
Thanks @sengi - I've moved this to its own script now. Do you see any problems with the approach before I give it a whirl on Integration?